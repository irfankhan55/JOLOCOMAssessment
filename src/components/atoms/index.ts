export { PrimaryTextField } from "./primary-input";
export { SecondaryInput } from "./secondary-Input";
export { PrimaryButton } from "./primary-button";
export { KeyboardAwareView } from "./keyboard-aware-view";
export { IconImage, IconSets, IconImageTouchable } from './icon-image';
