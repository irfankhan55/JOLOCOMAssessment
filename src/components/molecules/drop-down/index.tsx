import * as React from 'react';
import { ReactElement, useRef, useState, useLayoutEffect } from 'react';
import {
  FlatList,
  Text,
  TouchableOpacity,
  Modal,
  View,
  Animated,
  Easing
} from 'react-native';
import styles from './styles';
import Icon from 'react-native-vector-icons/FontAwesome';
import colors from '../../../res/colors';

type dataItem = {
  name: string; value: string
}
interface DropdownProps {
  defaultSelected: dataItem
  placeholder: string;
  data: dataItem[];
  onSelect: (item: { name: string; value: string }) => void;
  children?: React.ReactNode;
}

const Dropdown = (props: DropdownProps, {...rest}) => {
  const DropdownButton = useRef(undefined);
  const [visible, setVisible] = useState(false);
  const [selected, setSelected] = useState(props.defaultSelected || null);
  const [dropdownTop, setDropdownTop] = useState(0);
  const animation = useState(new Animated.Value(0))[0];

  useLayoutEffect(() => {
    animation.setValue(0);
    Animated.timing(animation, {
      toValue: visible ? 1 : 0,
      useNativeDriver: true,
      duration: 300,
      easing: Easing.linear,
    }).start();
  }, [visible])

  const IconRotationAngle = animation.interpolate({
    inputRange: [0, 1],
    outputRange: ["0deg", "180deg"]
  });

  const toggleDropdown = (): void => {
    openDropdown();
  };

  const openDropdown = (): void => {
    DropdownButton.current.measure((fx: number, fy: number, w: number, h: number, px: number, py: number) => {
      setDropdownTop(py + h);
      setVisible(true);
    });
  };

  const onItemPress = (item: dataItem): void => {
    setSelected(item);
    props.onSelect(item)
    setVisible(false);
  };

  const renderItem = ({ item }: { item: dataItem }) => (
    <TouchableOpacity style={styles.item} onPress={() => onItemPress(item)}>
      <Text style={styles.itemlabel}>{item.name}</Text>
    </TouchableOpacity>
  );

  const renderDropdown = (): ReactElement<any, any> => {
    return (
      <Modal visible={visible} transparent animationType='none'>
        <TouchableOpacity
          style={styles.overlay}
          onPress={() => setVisible(false)}
        >
          <View style={[styles.dropdown, { top: dropdownTop }]}>
            <FlatList
              data={props.data}
              renderItem={renderItem}
              keyExtractor={(item, index) => index.toString()}
            />
          </View>
        </TouchableOpacity>
      </Modal>
    );
  };

  return (
    <TouchableOpacity
      ref={DropdownButton}
      style={styles.button}
      onPress={toggleDropdown}
      {...rest}
    >
      {visible && renderDropdown()}
      <Text style={styles.buttonText}>
        {(selected && selected.name) || props.placeholder}
      </Text>
      <Animated.View style={[styles.icon, {transform: [{ rotate: IconRotationAngle }]}]}>
        <Icon name="chevron-down" size={20} color={colors.JOLOCOM_GRAY_TEXT}></Icon>
      </Animated.View>

    </TouchableOpacity>
  );
};

export { Dropdown };