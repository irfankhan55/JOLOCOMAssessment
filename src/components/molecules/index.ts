export { SideMenu } from "./side-menu";
export { Dropdown } from "./drop-down";
export { WorldCountryPicker } from "./country-picker";