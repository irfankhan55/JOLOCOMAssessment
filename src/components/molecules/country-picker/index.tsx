import * as React from "react";
import styles from "./styles";
import CountryPicker from 'react-native-country-picker-modal'
import { CountryCode, Country, Region} from './types'
import { View, Text, TextInputProps } from "react-native";
import Strings from "../../../res/i18n";
import { PrimaryTextField } from "../../atoms/primary-input";


const WorldCountryPicker = ({...rest}) => {
    const [countryCode, setCountryCode] = React.useState<CountryCode>(rest.countryCode)
    // const [region, setRegion] = React.useState<Region>('Europe') 
    const [country, setCountry] = React.useState<Country>(rest.country)
    const [withCountryNameButton] = React.useState<boolean>(
      true,
    )
    const [withFlag, setWithFlag] = React.useState<boolean>(true)
    const [withEmoji, setWithEmoji] = React.useState<boolean>(true)
    const [withFilter, setWithFilter] = React.useState<boolean>(true)
    const [withAlphaFilter, setWithAlphaFilter] = React.useState<boolean>(false)
    const [withCallingCode, setWithCallingCode] = React.useState<boolean>(false)
    const [withCloseButton, setwithCloseButton] = React.useState<boolean>(false)
    const onSelect = (country: Country) => {
      setCountryCode(country.cca2)
      setCountry(country)
    }  
    return (
      <View style={{width:'100%',flexDirection:'column'}}>
         <Text style={styles.titleStyle}>{Strings.whatCountryDoYouLiveIn}</Text>
          <CountryPicker
          {...{
            containerButtonStyle: styles.countryPickerStyle,
            modalProps:{ presentationStyle:'pageSheet'},
            renderCountryFilter: (props:TextInputProps) =>{
              return ( <PrimaryTextField
                          style={styles.inputStyle}
                          titleStyle={styles.countryPickerTitle}
                          title={Strings.country}
                          placeholder={Strings.selectCountry}
                          value={props.value}
                          onChangeText={(text:string) => props.onChangeText(text)}
                      />)
            },
            country,
            countryCode,
            withFilter,
            withFlag,
            withCountryNameButton,
            withCloseButton,
            withAlphaFilter,
            withCallingCode,
            withEmoji,
            onSelect,
            ...rest
          }}
        />
      </View>
    )};
export  { WorldCountryPicker }