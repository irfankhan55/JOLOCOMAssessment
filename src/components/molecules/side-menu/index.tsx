import * as React from 'react';
import { View, TouchableOpacity} from 'react-native';
import { DrawerContentComponentProps } from '@react-navigation/drawer';
import styles from './styles';
import { IconImage, IconSets } from '../../atoms';

type IconProps = { name: string; type: string };

type MenuItemProps = {
  isSelected: boolean;
  menuIcon: IconProps;
  onPress: () => void;
};

const MenuItem = ({ isSelected, menuIcon, onPress }: MenuItemProps) => {
  return (
    <TouchableOpacity
      style={[styles.menuItem, isSelected ? styles.menuItemSelected : {}]}
      onPress={onPress}
    >
      
        <IconImage
          iconSet={menuIcon.type}
          size={50}
          color={'green'}
          name={menuIcon.name}
        />
     
    </TouchableOpacity>
  );
};

type MenuIconsMap = {
  [key: string]: IconProps;
};

const menuIconsMap: MenuIconsMap = {
  HomeScreen: {
    name: 'logoWhite',
    type: IconSets.LOCAL_ICON
  },
  SettingsScreen: {
    name: 'check',
    type: IconSets.MATERIAL
  },
};
type SideMenuProps = DrawerContentComponentProps;

const SideMenu = (props: SideMenuProps) => {
  let selectedMenu = 'Dashboard';
  if (props && props.state && props.state.routes && props.state.index) {
    const selectedMenuName = props.state.routes[props.state.index].name;
    if (selectedMenuName) {
      selectedMenu = selectedMenuName;
    }
  }

  return (
    <View style={styles.box}>
      <View style={styles.menuContainer}>
        <View style={styles.menuItemsContainer}>  
          {props.state.routes.map(({ key, name }) => {
            return (
              <MenuItem
                onPress={() => {
                  props.navigation.navigate(name);
                }}
                key={key}
                menuIcon={menuIconsMap[name]}
                isSelected={selectedMenu === name}
              />
            );
          })}
        </View>
      </View>
    </View>
  );
};

export { SideMenu };
export type { SideMenuProps };
