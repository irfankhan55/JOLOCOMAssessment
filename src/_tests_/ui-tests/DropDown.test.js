import * as React from 'react';
import { Dropdown } from '../../components/molecules'
import TestRenderer, { act } from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';


describe('DropDown Testing', () => {
    const renderer = new ShallowRenderer();
    const placeholder = 'Select your Langauge'

    const mockCountries = [{ name: 'English (UK)', value: 'en' },
    { name: 'English (US)', value: 'en' },
    { name: 'French ', value: 'fr' },
    { name: 'Danish ', value: 'da' }]
  
    it('Render DropDown without Crashing', () => {
        const component = renderer.render(<Dropdown onSelect={() => { }} placeholder={placeholder} defaultSelected={mockCountries[0]} data={mockCountries} />);
        expect(component).toMatchSnapshot()
    });

    it('DropDown should have correct placeholder', () => {
        const testRenderer = TestRenderer.create(<Dropdown onSelect={() => { }} placeholder={placeholder} defaultSelected={mockCountries[0]} data={mockCountries} />);
        const testInstance = testRenderer.root;
        expect(testInstance.props.placeholder).toBe(placeholder);
    });

    it('DropDown should have correct data', () => {
        const testRenderer = TestRenderer.create(<Dropdown onSelect={() => { }} placeholder={placeholder} defaultSelected={mockCountries[0]} data={mockCountries} />);
        const testInstance = testRenderer.root;
        expect(testInstance.props.data.length).toBe(4);
    });

    it('DropDown should have correct default selected', () => {
        const testRenderer = TestRenderer.create(<Dropdown onSelect={() => { }} placeholder={placeholder} defaultSelected={mockCountries[0]} data={mockCountries} />);
        const testInstance = testRenderer.root;
        expect(testInstance.props.defaultSelected).toBe(mockCountries[0]);
    });
})          