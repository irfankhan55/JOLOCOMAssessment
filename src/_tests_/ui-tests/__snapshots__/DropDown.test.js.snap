// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown Testing Render DropDown without Crashing 1`] = `
<TouchableOpacity
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#101820",
      "borderColor": "#CCCCCC",
      "borderRadius": 4,
      "borderWidth": 1,
      "flexDirection": "row",
      "height": 50,
    }
  }
>
  <Text
    style={
      Object {
        "backgroundColor": "#101820",
        "color": "#999999",
        "flex": 1,
        "padding": 10,
        "textAlign": "left",
      }
    }
  >
    English (UK)
  </Text>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "color": "#CCCCCC",
          "marginRight": 10,
        },
        Object {
          "transform": Array [
            Object {
              "rotate": "0deg",
            },
          ],
        },
      ]
    }
  >
    <Icon
      allowFontScaling={false}
      color="#999999"
      name="chevron-down"
      size={20}
    />
  </ForwardRef(AnimatedComponentWrapper)>
</TouchableOpacity>
`;
